core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java

```diff
@@ -215,24 +224,48 @@ private void acquireConnectionAndExecute0(ClientRequestContext ctx, Endpoint end
         }
     }
 
-    private ProxyConfig getProxyConfig(SessionProtocol protocol, Endpoint endpoint) {
+    private void resolveProxyConfig(SessionProtocol protocol, Endpoint endpoint, ClientRequestContext ctx,
+                                    BiConsumer<@Nullable ProxyConfig, @Nullable Throwable> onComplete) {
         final ProxyConfig proxyConfig = factory.proxyConfigSelector().select(protocol, endpoint);
         requireNonNull(proxyConfig, "proxyConfig");
 
+        final ServiceRequestContext serviceCtx = ServiceRequestContext.currentOrNull();
+        final ProxiedAddresses capturedProxiedAddresses = serviceCtx == null ? null
+                                                          : serviceCtx.proxiedAddresses();
```

Comment:
Should we conditionally perform this block only when:
```java
proxyConfig.proxyType() == ProxyType.HAPROXY && 
((HAProxyConfig) proxyConfig).sourceAddress() == null
```

Reply:
I think so. 
But `maybeHAProxy()` can be executed in other thread because of `final Future<InetSocketAddress> resolveFuture = addressResolverGroup`. 

So, we should capture proxyAddress from `ServiceContext` in this thread.
Otherewise, the thread which perform `maybeHAProxy()` may get null even if the value is existed before. 

This is previous code which I extracted! 
```java
    private ProxyConfig getProxyConfig(SessionProtocol protocol, Endpoint endpoint) {
        if (proxyConfig.proxyType() == ProxyType.HAPROXY &&
            ((HAProxyConfig) proxyConfig).sourceAddress() == null) {
            final InetSocketAddress proxyAddress = proxyConfig.proxyAddress();
            assert proxyAddress != null;

            // use proxy information in context if available
            final ServiceRequestContext serviceCtx = ServiceRequestContext.currentOrNull();
            if (serviceCtx != null) {
                final ProxiedAddresses proxiedAddresses = serviceCtx.proxiedAddresses();
                return ProxyConfig.haproxy(proxyAddress, proxiedAddresses.sourceAddress());
            }
        }
}

```